File: test_vivid_scraper.kt
#!/usr/bin/env kotlin

// Standalone test script for VividSeats scraping
// Run with: kotlinc -script test_vivid_scraper.kt

@file:Repository("https://repo1.maven.org/maven2")
@file:DependsOn("com.squareup.okhttp3:okhttp:4.11.0")
@file:DependsOn("org.jsoup:jsoup:1.15.3")
@file:DependsOn("org.json:json:20230227")

import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import org.json.JSONObject
import org.json.JSONArray

fun main() {
    println("🔍 VividSeats World Cup Scraper Test")
    println("=" * 60)

    val client = OkHttpClient()
    val url = "https://www.vividseats.com/world-cup-soccer-tickets--sports-soccer/performer/944"

    try {
        println("\n📡 Fetching $url...")
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1")
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            println("❌ HTTP Error: ${response.code}")
            return
        }

        val html = response.body?.string() ?: run {
            println("❌ Empty response body")
            return
        }

        println("✓ Got HTML response (${html.length} bytes)")

        // Try JSON extraction
        println("\n📋 Attempting JSON extraction...")
        val jsonListings = extractFromJson(html)
        if (jsonListings.isNotEmpty()) {
            println("✓ JSON extraction SUCCESS: Found ${jsonListings.size} games")
            jsonListings.take(5).forEach { (name, venue, price, date) ->
                println("  - $name @ $venue on $date ($$$price)")
            }
            if (jsonListings.size > 5) {
                println("  ... and ${jsonListings.size - 5} more")
            }
            return
        }

        println("✗ JSON extraction failed or empty, trying HTML fallback...")

        // Try HTML parsing
        println("\n🔎 Attempting HTML parsing...")
        val htmlListings = parseFromHtml(html)
        println("✓ HTML parsing complete: Found ${htmlListings.size} games")
        htmlListings.take(5).forEach { (name, venue, price, date) ->
            println("  - $name @ $venue on $date ($$$price)")
        }
        if (htmlListings.size > 5) {
            println("  ... and ${htmlListings.size - 5} more")
        }

    } catch (e: Exception) {
        println("❌ Error: ${e.message}")
        e.printStackTrace()
    }
}

data class Listing(val name: String, val venue: String, val price: Int, val date: String)

fun extractFromJson(html: String): List<Listing> {
    return try {
        val doc = Jsoup.parse(html)
        val script = doc.select("script#__NEXT_DATA__").firstOrNull()?.data() ?: run {
            println("  ! __NEXT_DATA__ script not found")
            return emptyList()
        }

        println("  ✓ Found __NEXT_DATA__ script (${script.length} bytes)")

        val events = JSONObject(script)
            .getJSONObject("props")
            .getJSONObject("pageProps")
            .getJSONObject("initialData")
            .getJSONArray("events")

        println("  ✓ Parsed JSON structure, found ${events.length()} events")

        val results = mutableListOf<Listing>()
        var skipped = 0

        for (i in 0 until events.length()) {
            val event = events.getJSONObject(i)
            val venueObj = event.optJSONObject("venue")
            val venueName = venueObj?.optString("name") ?: "Unknown"
            val venueCity = venueObj?.optString("city") ?: "Unknown"

            // Basic stadium matching (simplified from actual)
            val stadiumMatch = when {
                venueName.contains("Toronto", ignoreCase = true) -> "Toronto"
                venueName.contains("Vancouver", ignoreCase = true) -> "Vancouver"
                venueName.contains("Mexico", ignoreCase = true) -> "Mexico City"
                venueCity.contains("Guadalajara", ignoreCase = true) -> "Guadalajara"
                venueCity.contains("Monterrey", ignoreCase = true) -> "Monterrey"
                venueName.contains("Atlanta", ignoreCase = true) -> "Atlanta"
                venueName.contains("Boston", ignoreCase = true) -> "Boston"
                venueName.contains("Dallas", ignoreCase = true) -> "Dallas"
                venueName.contains("Houston", ignoreCase = true) -> "Houston"
                venueName.contains("Kansas", ignoreCase = true) -> "Kansas City"
                venueName.contains("Los Angeles", ignoreCase = true) || venueName.contains("SoFi", ignoreCase = true) -> "Los Angeles"
                venueName.contains("Miami", ignoreCase = true) -> "Miami"
                venueCity.contains("East Rutherford", ignoreCase = true) || venueName.contains("MetLife", ignoreCase = true) -> "New York"
                venueName.contains("Philadelphia", ignoreCase = true) -> "Philadelphia"
                venueName.contains("San Francisco", ignoreCase = true) -> "San Francisco"
                venueName.contains("Seattle", ignoreCase = true) -> "Seattle"
                venueCity.contains("Washington", ignoreCase = true) -> "Washington D.C."
                venueName.contains("Orlando", ignoreCase = true) -> "Orlando"
                venueName.contains("Charlotte", ignoreCase = true) -> "Charlotte"
                venueName.contains("Chicago", ignoreCase = true) -> "Chicago"
                venueCity.contains("Las Vegas", ignoreCase = true) || venueName.contains("Allegiant", ignoreCase = true) -> "Las Vegas"
                venueName.contains("Phoenix", ignoreCase = true) -> "Phoenix"
                venueName.contains("Denver", ignoreCase = true) -> "Denver"
                else -> null
            }

            if (stadiumMatch == null) {
                println("    ! Could not match stadium: $venueName, $venueCity")
                skipped++
                continue
            }

            val name = event.optString("name", "World Cup Match")
                .replace(Regex("(?i)tickets$"), "").trim()

            results.add(Listing(
                name = name,
                venue = stadiumMatch,
                price = event.optInt("minPrice", 0),
                date = event.optString("date", "")
            ))
        }

        println("  ✓ JSON parsed: ${results.size} matched, $skipped skipped")
        results
    } catch (e: Exception) {
        println("  ✗ JSON extraction error: ${e.message}")
        e.printStackTrace()
        emptyList()
    }
}

fun parseFromHtml(html: String): List<Listing> {
    println("  Starting HTML parsing...")
    val doc = Jsoup.parse(html)
    val elements = doc.select("[data-testid^=production-listing]")

    println("  ✓ Found ${elements.size} HTML elements with [data-testid^=production-listing]")

    val results = mutableListOf<Listing>()
    var skipped = 0

    elements.forEach { el ->
        val text = el.text()

        // Simplified stadium detection
        val stadium = when {
            text.contains("Toronto", ignoreCase = true) -> "Toronto"
            text.contains("Vancouver", ignoreCase = true) -> "Vancouver"
            text.contains("Mexico", ignoreCase = true) -> "Mexico City"
            text.contains("Guadalajara", ignoreCase = true) -> "Guadalajara"
            text.contains("Monterrey", ignoreCase = true) -> "Monterrey"
            text.contains("Atlanta", ignoreCase = true) -> "Atlanta"
            text.contains("Boston", ignoreCase = true) || text.contains("Gillette", ignoreCase = true) -> "Boston"
            text.contains("Dallas", ignoreCase = true) -> "Dallas"
            text.contains("Houston", ignoreCase = true) -> "Houston"
            text.contains("Kansas", ignoreCase = true) -> "Kansas City"
            text.contains("Los Angeles", ignoreCase = true) || text.contains("SoFi", ignoreCase = true) -> "Los Angeles"
            text.contains("Miami", ignoreCase = true) -> "Miami"
            text.contains("MetLife", ignoreCase = true) || text.contains("New York", ignoreCase = true) -> "New York"
            text.contains("Philadelphia", ignoreCase = true) -> "Philadelphia"
            text.contains("San Francisco", ignoreCase = true) -> "San Francisco"
            text.contains("Seattle", ignoreCase = true) -> "Seattle"
            text.contains("Washington", ignoreCase = true) -> "Washington D.C."
            text.contains("Orlando", ignoreCase = true) -> "Orlando"
            text.contains("Charlotte", ignoreCase = true) -> "Charlotte"
            text.contains("Chicago", ignoreCase = true) -> "Chicago"
            text.contains("Vegas", ignoreCase = true) -> "Las Vegas"
            text.contains("Phoenix", ignoreCase = true) -> "Phoenix"
            text.contains("Denver", ignoreCase = true) -> "Denver"
            else -> null
        }

        if (stadium == null) {
            println("    ! Skipped element (no stadium match): ${text.take(60)}...")
            skipped++
            return@forEach
        }

        val price = Regex("\\$(\\d[\\d,]*)").find(text)?.groupValues?.get(1)?.replace(",", "")?.toIntOrNull() ?: 0
        val date = Regex("\\b(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\\s+\\d{1,2}", RegexOption.IGNORE_CASE).find(text)?.value ?: ""

        results.add(Listing(text.take(40), stadium, price, date))
    }

    println("  ✓ HTML parsed: ${results.size} matched, $skipped skipped")
    return results
}

operator fun String.times(count: Int) = this.repeat(count)

main()

File: test_vivid_scraper.py
#!/usr/bin/env python3
"""
Standalone test script for VividSeats World Cup scraping
Run with: python3 test_vivid_scraper.py
"""

import requests
import json
from bs4 import BeautifulSoup
import re
import sys

STADIUMS = [
    "Toronto", "Vancouver", "Mexico City", "Guadalajara", "Monterrey",
    "Atlanta", "Boston", "Dallas", "Houston", "Kansas City", "Los Angeles",
    "Miami", "New York", "Philadelphia", "San Francisco", "Seattle",
    "Washington", "Orlando", "Charlotte", "Chicago", "Las Vegas",
    "Phoenix", "Denver"
]

VENUES = {
    "Toronto": ["toronto", "bmo"],
    "Vancouver": ["vancouver", "bc place"],
    "Mexico City": ["mexico city", "azteca", "cdmx"],
    "Guadalajara": ["guadalajara", "akron"],
    "Monterrey": ["monterrey", "bbva"],
    "Atlanta": ["atlanta", "mercedes"],
    "Boston": ["boston", "gillette"],
    "Dallas": ["dallas", "at&t"],
    "Houston": ["houston", "nrg"],
    "Kansas City": ["kansas city", "arrowhead"],
    "Los Angeles": ["los angeles", "sofi"],
    "Miami": ["miami", "hard rock"],
    "New York": ["new york", "metlife"],
    "Philadelphia": ["philadelphia", "lincoln"],
    "San Francisco": ["san francisco", "levi"],
    "Seattle": ["seattle", "lumen"],
    "Washington": ["washington", "audi field"],
    "Orlando": ["orlando", "camping"],
    "Charlotte": ["charlotte"],
    "Chicago": ["chicago"],
    "Las Vegas": ["vegas", "allegiant"],
    "Phoenix": ["phoenix"],
    "Denver": ["denver"]
}

def match_stadium(text):
    """Match stadium name from text"""
    text_lower = text.lower()

    # Try venue names first
    for stadium, venues in VENUES.items():
        for venue in venues:
            if venue in text_lower:
                return stadium

    # Try stadium names
    for stadium in STADIUMS:
        if stadium.lower() in text_lower:
            return stadium

    return None

def extract_from_json(html):
    """Extract listings from __NEXT_DATA__ script"""
    soup = BeautifulSoup(html, 'html.parser')
    script = soup.find('script', {'id': '__NEXT_DATA__'})

    if not script:
        print("  ! __NEXT_DATA__ script not found")
        return []

    print(f"  ✓ Found __NEXT_DATA__ script ({len(script.string)} bytes)")

    try:
        data = json.loads(script.string)
        events = data['props']['pageProps']['initialData']['events']
        print(f"  ✓ Parsed JSON structure, found {len(events)} events")
    except Exception as e:
        print(f"  ✗ JSON structure error: {e}")
        return []

    results = []
    skipped = 0

    for event in events:
        venue = event.get('venue', {})
        venue_name = venue.get('name', 'Unknown')
        venue_city = venue.get('city', 'Unknown')
        venue_text = f"{venue_name} {venue_city}"

        stadium = match_stadium(venue_text)
        if not stadium:
            print(f"    ! Could not match stadium: '{venue_text}'")
            skipped += 1
            continue

        name = event.get('name', 'World Cup Match').replace('Tickets', '').strip()
        price = event.get('minPrice', 0)
        date = event.get('date', '')

        results.append({
            'name': name,
            'stadium': stadium,
            'price': price,
            'date': date
        })

    print(f"  ✓ JSON parsed: {len(results)} matched, {skipped} skipped")
    return results

def parse_from_html(html):
    """Parse listings from HTML elements"""
    soup = BeautifulSoup(html, 'html.parser')
    elements = soup.find_all(attrs={'data-testid': re.compile(r'^production-listing')})

    print(f"  ✓ Found {len(elements)} HTML elements with [data-testid^=production-listing]")

    results = []
    skipped = 0

    for el in elements:
        text = el.get_text()
        stadium = match_stadium(text)

        if not stadium:
            print(f"    ! Skipped element (no stadium match): {text[:60]}...")
            skipped += 1
            continue

        # Extract price
        price_match = re.search(r'\$(\d[\d,]*)', text)
        price = int(price_match.group(1).replace(',', '')) if price_match else 0

        # Extract date
        date_match = re.search(r'\b(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+\d{1,2}', text, re.IGNORECASE)
        date = date_match.group(0) if date_match else ''

        results.append({
            'name': text[:40],
            'stadium': stadium,
            'price': price,
            'date': date
        })

    print(f"  ✓ HTML parsed: {len(results)} matched, {skipped} skipped")
    return results

def main():
    print("🔍 VividSeats World Cup Scraper Test")
    print("=" * 60)

    url = "https://www.vividseats.com/world-cup-soccer-tickets--sports-soccer/performer/944"
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'Referer': 'https://www.vividseats.com/',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1'
    }

    try:
        print(f"\n📡 Fetching {url}...")
        response = requests.get(url, headers=headers, timeout=10)

        if response.status_code != 200:
            print(f"❌ HTTP Error: {response.status_code}")
            return

        html = response.text
        print(f"✓ Got HTML response ({len(html)} bytes)")

        # Try JSON extraction
        print("\n📋 Attempting JSON extraction...")
        json_listings = extract_from_json(html)

        if json_listings:
            print(f"✓ JSON extraction SUCCESS: Found {len(json_listings)} games")
            for listing in json_listings[:5]:
                print(f"  - {listing['name']} @ {listing['stadium']} on {listing['date']} (${listing['price']})")
            if len(json_listings) > 5:
                print(f"  ... and {len(json_listings) - 5} more")
            return

        # Try HTML parsing
        print("\n🔎 Attempting HTML parsing...")
        html_listings = parse_from_html(html)

        print(f"✓ HTML parsing complete: Found {len(html_listings)} games")
        for listing in html_listings[:5]:
            print(f"  - {listing['name']} @ {listing['stadium']} on {listing['date']} (${listing['price']})")
        if len(html_listings) > 5:
            print(f"  ... and {len(html_listings) - 5} more")

    except requests.exceptions.RequestException as e:
        print(f"❌ Network error: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == '__main__':
    main()
